### ISSUE TITLE
Certificate generation bash script for samples

### ISSUE THREAD

**Issue by aurashn** [Azure/azure-iot-sdk-c#580]
- **OS and version used:** <VERSION> 
 Ubuntu 16.04
- **SDK version used:** <VERSION> 
azure-iot-sdk-c, 1.2.8

# Description of the issue:

I would like to test provisioning a device to a group enrollment. I have used the certGen.sh in the tools under the c sdk to create my certificate and verified it with my DPS. Next, I added the device leaf cert and key in the file custom_hsm_sample. After building the library, I used the cmake options specified in the docs to run cmake and build.

`sudo cmake -Duse_prov_client:BOOL=ON -Dhsm_custom_lib="/home/aurash/Azure-c-Iot/azure-iot-sdk-c/provisioning_client/samples/custom_hsm_example/cmake/libcustom_hsm_example.a" "/home/aurash/Azure-c-Iot/azure-iot-sdk-c/"`

Next, I am under the impression that I should be able to run the prov_dev_client_sample.c with my url and scope to provision my device to the iot hub using the DPS and the enrollment group I created.

# Code sample exhibiting the issue:

```


# Console log of the issue:
<!-- Please share as much logs as posible, that will help debugging -->
<!-- Don't forget to remove any connection string information! -->

Provisioning API Version: 1.2.8
Iothub API Version: 1.2.8
Error: Time:Tue Jul 31 17:32:14 2018 File:/home/aurash/Azure-c-Iot/azure-iot-sdk-c/provisioning_client/src/prov_transport_mqtt_common.c Func:mqtt_operation_complete_callback Line:201 Connection Not Accepted: 0x5: Not Authorized
Error: Time:Tue Jul 31 17:32:14 2018 File:/home/aurash/Azure-c-Iot/azure-iot-sdk-c/provisioning_client/src/prov_transport_mqtt_common.c Func:mqtt_error_callback Line:134 MQTT communication error
Error: Time:Tue Jul 31 17:32:14 2018 File:/home/aurash/Azure-c-Iot/azure-iot-sdk-c/provisioning_client/src/prov_device_ll_client.c Func:on_transport_registration_data Line:455 Failure retrieving data from the provisioning service
Failure encountered on registration!
registration failed!
Press any enter to continue:

**Comment by jspaith** [comment]
In your initial post before the edit, you said that this wasn't working when using the x509 sample.  Is this still the case or is it going and problems are limited to custom hsm?  

1 - 
In general, if you could add this code to whatever client code you're using (you may need to remove the _LL_ and replace the DeviceClient with just a Client depending on which API set you're using) and give us logs that could provide more data.

```c
        bool traceOn = true;
        IoTHubDeviceClient_LL_SetOption(device_ll_handle, OPTION_LOG_TRACE, &traceOn);
```

2 - Another interesting data point here is going to be whether this works when doing a CACertificate, never mind DPS.  This is assuming your original x509 doesn't work either, it's not only custom HSM, otherwise ignore below.

This is basically from [here](https://github.com/Azure/azure-iot-sdk-c/blob/master/tools/CACertificates/CACertificateOverview.md) steps like in your IoT Hub with `Next, go to Azure IoT Hub and navigate to Certificates` and then `Step 3 - Proof of Possession` instructions.  That I think uses a lot of same plumbing on back end but a few fewer moving pieces so could be easier to debug.

In this, *you'd use the original x509 sample and not custom hsm* and again having logs might tell us more.

**Comment by aurashn** [comment]
I ran the x509 sample as you mentioned above by creating a new root certificate with certGen.sh and using the IotHub to verify the certificate than add the device under the device explorer. It compiles and builds, yet upon running I am receiving the similar errors. Also, I activated the log as you mentioned.

Creating IoTHub handle
Sending message 1 to IoTHub
Sending message 2 to IoTHub
Sending message 3 to IoTHub
Sending message 4 to IoTHub
Sending message 5 to IoTHub
-> 11:00:35 CONNECT | VER: 4 | KEEPALIVE: 240 | FLAGS: 128 | USERNAME: FleetIoT.azure-devices.net/mydevice/?api-version=2017-11-08-preview&DeviceClientType=iothubclient%2f1.2.8%20(native%3b%20Linux%3b%20x86_64) | CLEAN: 0
<- 11:00:35 CONNACK | SESSION_PRESENT: false | RETURN_CODE: 0x5
Error: Time:Wed Aug  1 11:00:35 2018 File:/home/aurash/Azure-c-Iot/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:mqtt_operation_complete_callback Line:1567 Connection Not Accepted: 0x5: Not Authorized
-> 11:00:35 CONNECT | VER: 4 | KEEPALIVE: 240 | FLAGS: 128 | USERNAME: FleetIoT.azure-devices.net/mydevice/?api-version=2017-11-08-preview&DeviceClientType=iothubclient%2f1.2.8%20(native%3b%20Linux%3b%20x86_64) | CLEAN: 0
<- 11:00:35 CONNACK | SESSION_PRESENT: false | RETURN_CODE: 0x5
Error: Time:Wed Aug  1 11:00:35 2018 File:/home/aurash/Azure-c-Iot/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:mqtt_operation_complete_callback Line:1567 Connection Not Accepted: 0x5: Not Authorized
-> 11:00:37 CONNECT | VER: 4 | KEEPALIVE: 240 | FLAGS: 128 | USERNAME: FleetIoT.azure-devices.net/mydevice/?api-version=2017-11-08-preview&DeviceClientType=iothubclient%2f1.2.8%20(native%3b%20Linux%3b%20x86_64) | CLEAN: 0
<- 11:00:37 CONNACK | SESSION_PRESENT: false | RETURN_CODE: 0x5
Error: Time:Wed Aug  1 11:00:37 2018 File:/home/aurash/Azure-c-Iot/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:mqtt_operation_complete_callback Line:1567 Connection Not Accepted: 0x5: Not Authorized
-> 11:00:41 CONNECT | VER: 4 | KEEPALIVE: 240 | FLAGS: 128 | USERNAME: FleetIoT.azure-devices.net/mydevice/?api-version=2017-11-08-preview&DeviceClientType=iothubclient%2f1.2.8%20(native%3b%20Linux%3b%20x86_64) | CLEAN: 0
<- 11:00:41 CONNACK | SESSION_PRESENT: false | RETURN_CODE: 0x5
Error: Time:Wed Aug  1 11:00:41 2018 File:/home/aurash/Azure-c-Iot/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:mqtt_operation_complete_callback Line:1567 Connection Not Accepted: 0x5: Not Authorized

**Comment by jspaith** [comment]
It looks like something is wrong with the bash based certificate generation, since I hit issues here too.  My gut is the script is misconfiguring one of the many openssl options and that's getting things off.

For a workaround, I've verified that following the PowerShell instructions do indeed give a perfectly good cert.  I'm using the x509 sample / proof of possession sample, not DPS.  Since neither the PowerShell nor bash are hardened for production certs, but just for development purposes, if you have a Win box this should unblock you for now.

That said, I'm changing this to a bug since the bash script should work, too.  Sorry you've run into this.

**Comment by lakshmisivareddy** [comment]
Hi @aurashn 
/azure-iot-sdk-c/provisioning_client/src/prov_transport_mqtt_common.c Func:mqtt_operation_complete_callback Line:201 Connection Not Accepted: 0x5: Not Authorized
this error will occur only  when the device certificate chain that your passing not signed by the same CA that you uploaded in portal 
i also observed this error when i run x509 sample uploading CA certificate in portal 
c/iothub_client/src/iothubtransport_mqtt_common.c Func:mqtt_operation_complete_callback Line:1567 Connection Not Accepted: 0x5: Not Authorized

if possible can you share the wire shark log so that we can see why SSL  handshake is failing

**Comment by alexsavio** [comment]
I am hitting the same issue. I think there is something wrong with the bash script, since we also tried with the PowerShell version and it worked.

**Comment by alexsavio** [comment]
The issue seems to be that the instructions lead the user to upload the root CA certificate to IoTHub, however the device certificates are generated with the intermediate certificate. Also the full-chain certificate that is mentioned to be generated in the end is not valid, I think it needs a private key somewhere. I am slowly trying to debug this.

**Comment by alexsavio** [comment]
Found the issue: 

In the end I do the following instead of what is suggested in the Bash/IoT Leaf Device section of the CACertificateOverview.md.

```bash
openssl pkcs12 -in new-device.cert.pfx -out new-device.cert.nodes.pem -nodes
cat new-device.cert.nodes.pem azure-iot-test-only.intermediate.cert.pem > new-device-full-chain.cert.pem
```

**Comment by RobinTTY** [comment]
As others said the bash script seems to contain a bug, I wasn't able to create a valid device certificate with it, when trying to connect with the certificate, my IoT Hub logs "None of the intermediate CAs were found in the list of allowed CAs". The PowerShell script is working fine. It might be good to make a note in CACertificateOverview.md that step 4 doesn't work with the bash script as is currently.

**Comment by danewalton** [comment]
PR Opened
https://github.com/Azure/azure-iot-sdk-c/pull/1230

**Comment by danewalton** [comment]
PR has been merged to master. I confirmed that generating certs from `certGen.sh` works with iothub_x509 sample and provisioning with custom hsm. If anyone has any other issues please feel free to open another issue.

**Comment by az-iot-builder-01** [comment]
@jspaith, @aurashn, @lakshmisivareddy, @alexsavio, @RobinTTY, @danewalton, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
